### ISSUE TITLE
FEATURE REQUEST: Add kubelet instance config

### ISSUE THREAD

**Issue by HirazawaUi** [kubernetes/kubeadm#3042]
## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

## Anything else we need to know?

This issue wants to remove container runtime interface (CRI) socket annotations from Node objects in Kubernetes and no longer added `--container-runtime-endpoint` args in `kubeadm-flags.env` ( `--container-runtime-endpoint` args deprecated in kubelet), this annotation and args are used to specify the CRI socket endpoint on each node that the kubelet uses to communicate with the container runtime.

Add `kubelet-instance-config.yml` locally to store container runtime interface (CRI) sockets, and in `kubeadm init`, `join` and `upgrade`, override the global kubelet configuration through `kubelet-instance-config.yml`.

More detailed design at: https://hackmd.io/@r2Yq-PVoRcK2OaGypnwPMg/rJFChACCp/edit

**Comment by neolit123** [comment]
@HirazawaUi we need a KEP for this, ket's see what @pacoxu can comment about the previous work done on this.

@pacoxu can you please post links to your previous KEP work.?
also didn't we already have a tracking issue here in k/kubeadm?

maybe we can close KEP PRs and let @HirazawaUi to take over.

**Comment by HirazawaUi** [comment]
> also are you willing to work on this @HirazawaUi there are a lot of things to update around this change across multiple releases. and i don't think me or anyone else will have a lot of time for it.
> 
> * KEP design how is it going to work exactly, do we need a feature gate
> * k8s.io docs update
> * k/k code update
> * do we need a e2e test in k/kubeadm
> 
> too many questions

Yes, I'll try to finish it.

**Comment by HirazawaUi** [comment]
@pacoxu What do you think about this? I would love to hear your comments :)

**Comment by pacoxu** [comment]
The KEP 
- https://github.com/kubernetes/enhancements/issues/3929
- https://github.com/kubernetes/enhancements/pull/3930

My current proposal would be something like  https://github.com/kubernetes/enhancements/pull/3930#issuecomment-1931657825
> > i meant a generic patches approach where a local config file overrides the global config downloaded from the cm. not the --patches feature.
> > 
> > this local file must be stored somewhere. perhaps in the same dir as config.yaml, but called config-instance.yaml. IMO it has a number of tricky aspects that need to be covered in the design doc for init, join, upgrade.
> > 

> Above is @neolit123 your proposal in comment here https://github.com/kubernetes/enhancements/pull/3930/files#r1177682829.
> 
> I prefer this solution.
> 
> BTW, /var/lib/kubelet/kubeadm-flags.env will only have --container-runtime-endpoint later. I would like to make this cri-socket-annotation and runtime endpoint configuration to config-instance.yaml under /var/lib/kubelet.

**Comment by HirazawaUi** [comment]
Thanks, I have missed this comment, I will resurrect kep as soon as possible.

**Comment by neolit123** [comment]
note if this ever becomes part of kubeadm, we can close this ticket:
https://github.com/kubernetes/kubeadm/issues/1924

**Comment by neolit123** [comment]
@HirazawaUi i added some TODOs a tracking in the issue description here.
please keep it updated with links `- ....`

**Comment by HirazawaUi** [comment]
> @HirazawaUi i added some TODOs a tracking in the issue description here. please keep it updated with links `- ....`

OK, thank you so much for your help.

**Comment by neolit123** [comment]
@HirazawaUi @pacoxu 

i think i found one problem wih the feature gate.
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/postupgrade.go#L138

here we read the value from the env file, but we also need to migrate users away from using the flag --container-runtime for the kubelet. so on upgrade we need to mutate the env file to no longer have the --container-runtime flag. the problem here is that if the kubelet receives the flag --container-runtime then it will take precedence over the value passed from the kubelet config.yaml and the config yaml will be ignored.

this is actually something we might have to backport as a fix for 1.32. in the alpha.

**Comment by pacoxu** [comment]
Probably, we can keep it as in v1.32, as the value in env file and instance file are the same.
- v1.33+: mutate the env file during upgrade
- v1.34 make it beta?

**Comment by neolit123** [comment]
i think i agree to not backport the fix for this alpha feature.
i guess the problem is that as long as --container-runtime is in the env file it means the feature gate doesn't work.
for new clusters however init/join would not have the --container-runtime flag in the env file i think @HirazawaUi to confirm.

**Comment by HirazawaUi** [comment]
> here we read the value from the env file, but we also need to migrate users away from using the flag --container-runtime for the kubelet. so on upgrade we need to mutate the env file to no longer have the --container-runtime flag. the problem here is that if the kubelet receives the flag --container-runtime then it will take precedence over the value passed from the kubelet config.yaml and the config yaml will be ignored.

Yes, that is indeed a problem.






> for new clusters however init/join would not have the --container-runtime flag in the env file i think @HirazawaUi to confirm.

That's right.


Given that we have never mentioned this (modifying an existing env file) in the KEP, I agree with @pacoxu's view that we should keep it as is in 1.32 and move it to beta in 1.34, WDYT?

I will update the KEP later to add this description and revise it according to the implemented PR (I had planned to do this last week, but it was delayed due to being reinfected with COVID-19).

**Comment by neolit123** [comment]
> 1.32 and move it to beta in 1.34, WDYT?

1.34 SGTM, but this issue is something we can fix in 1.33 as well and still keep it alpha.
we can then check in the e2e test the flag is correctly removed.

> I will update the KEP later to add this description and revise it according to the implemented PR (I had planned to do this last week, but it was delayed due to being reinfected with COVID-19).

ok, take care of your self.

**Comment by neolit123** [comment]
> > 1.32 and move it to beta in 1.34, WDYT?
> 
> 1.34 SGTM, but this issue is something we can fix in 1.33 as well and still keep it alpha. we can then check in the e2e test the flag is correctly removed.
> 

@HirazawaUi 
something else that we are missing:
- if the FG is enabled before upgrade the annotation must be removed from the Node objects of all nodes.
- it should be a specific cleanup present in the post-upgrade area of kubeadm until we graduate the FG to GA.

https://github.com/kubernetes/kubernetes/pull/128031
did not have it.

**Comment by HirazawaUi** [comment]
> i added these 2 things as TODO in the OP here.

All right.

**Comment by neolit123** [comment]
@HirazawaUi would you have bandwidth to progress this in 1.34?

**Comment by HirazawaUi** [comment]
> [@HirazawaUi](https://github.com/HirazawaUi) would you have bandwidth to progress this in 1.34?

Yes, I will complete it this week.

**Comment by neolit123** [comment]
note: for GA we should update these pages too. IIRC, they do reference the pre-FG way, so given the behavior will be locked on enabled we should document it:
- https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/ (we can mention the instance-config.yaml)
- https://kubernetes.io/docs/reference/setup-tools/kubeadm/implementation-details/ (cover the new instance-config, remove the annotation mentions)
- https://kubernetes.io/docs/reference/labels-annotations-taints/ (remove the annotation mentions)

**Comment by HirazawaUi** [comment]
> [@HirazawaUi](https://github.com/HirazawaUi)
> 
> after the beta pr merges we should update the 'kubeadm init' website page and flip the kinder e2e to test fg=disabled

Yes, this is on my plan. I expect to tackle it tomorrow or the day after.

**Comment by neolit123** [comment]
@HirazawaUi please open the k/website pr to update the feature gate graduation. branch is dev-1.34. today is the deadline for opening docs prs.

**Comment by HirazawaUi** [comment]
As @neolit123 pointed out, this is a known issue and we have fixed it.
